### DISCUSSION TITLE
Wave settings from state space realization

### DISCUSSION THREAD

**Discussion by QusayHawari** [OpenFAST/openfast#1053] [Q&A]
Hi all,

Maybe this is a simple question. Suppose I am linearizing a turbine model and wanted to include the wave excitation and radiation effects by setting ExctnMod = RdtnMod = 2 to be included as state space models. Is there a setting in OpenFAST where I can change some of the parameters such as wave height to be updated in the state space models? I am not sure I understand what all the states in the .ss and .ssexctn represent. Is there a way where I can find out how to edit wave settings from the state space models? 

Thanks in advance,
Qusay.

**Comment by jjonkman** [comment]
Der @QusayHawari,

Typically, the OpenFAST model is linearized in a still water condition in order to ensure that the offshore wind turbine can reach a steady state (or periodic steady state) condition before linearizing.  In the linearized representation of the model output from OpenFAST, the states, inputs, and outputs represent perturbations about the operating point.  For the wave disturbance input, then, the operating point wave elevation is zero and the perturbation of the wave disturbance input is the actual wave elevation.

Regarding the state-space implementations of wave radiation and wave excitation, the states capture the frequency dependence and memory effects. Hopefully the following papers clarify: https://www.nrel.gov/docs/fy13osti/58099.pdf and https://www.nrel.gov/docs/fy19osti/71865.pdf.

To derive the state-space wave-radiation model, you can use the SS_Fitting preprocessor: https://www.nrel.gov/wind/nwtc/ss-fitting.html.

To derive the state-space wave-excitation model, a similar preprocessor has been developed named SS_Excitation_Fitting, as discussed here: https://github.com/OpenFAST/openfast/issues/397.

Best regards,

**Reply by QusayHawari** [reply]
Hi Jason,

I wanted to generate an *ssexctn file for the WindCrete spar 15MW floater, but I could not find the MATLAB fitting toolbox for it. 
 You mentioned in #397  to contact Alan, but Alan is retired now as you outlined here https://forums.nrel.gov/t/wamit-ss-file/1358/12 

Is there a way I can reach out to Alan, or you can send me a beta version of the toolbox?

Thanks,
Qusay

**Reply by jjonkman** [reply]
Dear @QusayHawari,

If you provide your e-mail address, I can send you the alpha release of SS_Excitation_Fitting via WeTransfer.

Best regards,

**Reply by QusayHawari** [reply]
Hi,

q.hawari@lboro.ac.uk

Thanks,
Qusay

**Reply by jjonkman** [reply]
I've sent you the alpha release of SS_Excitation_Fitting via WeTransfer.

**Reply by QusayHawari** [reply]
Received. 

Thank you very much!

Qusay

**Reply by KANG20240612** [reply]
Dear @jjonkman 
Could you send me the alpha release of SS_Excitation_Fitting?
My e-mail: jianfei_kang@usts.edu.cn

Thanks,
Jianfei Kang

**Reply by jjonkman** [reply]
I've sent you the alpha release of SS_Excitation_Fitting via WeTransfer.

**Comment by QusayHawari** [comment]
Hi,


Thank you for the useful material. 

I understand that the linearizations are always performed about the operating points. I completely forgot that the state space models have an input for the wave elevation perturbations about the 0 (operating point).  

The .ss and .ssexctn state space files for the 15 MW floater have already been compiled and are available here https://github.com/IEAWindTask37/IEA-15-240-RWT/tree/develop/OpenFAST/IEA-15-240-RWT-UMaineSemi/HydroData .

I was just not sure of the logical operation when adding the wave excitation and radiation effects into the linearization. 

-------------------------------------------

Please correct me if I am wrong. 

Wave excitation state space models are used to depict the behaviour of the platform DOFs(6 max) in wave conditions assuming a fixed platform. 

The radiation state space models basically show the opposite effect, i.e. checking the platform DOFs in still water under a moving platform. 

Assuming I am using a linearized model and I inserted 5m as an input wave elevation disturbance. From that disturbance we can see how the platform DOFs behave. Not only that, but we also use those platform DOF behaviours as inputs to the radiation state space model to see how those DOFs  behave under the movements of the platform. Adding both platform effects together will get the full response from wave excitation and radiation. Is that accurate?


Thanks,
Qusay.

**Reply by jjonkman** [reply]
Dear @QusayHawari,

I agree with your description.  I would say that the wave-excitation loads from the potential-flow model capture the applied hydrodynamic loading (forces, moments) on the body when the body is fixed in position and incident waves are present.  And the wave-radiation loads from the potential-flow model capture the applied hydrodynamic loading (forces, moments) on the body when no incident waves are present, but waves are generated by the motion of the body.

Best regards,

**Reply by QusayHawari** [reply]
Hi,

Thank you for your reply. 

Just to check (for validating linear with nonlinear) that a nonlinear model with the following settings: _WaveMod =1, WaveHs = 5, PotMod = ExctnMod = RdtnMod= 1_ will correspond to a linearized model with the settings: _WaveMod = 0, WaveHs = N/A, PotMod = 1, ExctnMod = RdtnMod= 2_ with a disturbance of 5m as the wave elevation input to the state space models.

I tried testing the nonlinear response by setting ExctnMod = RdtnMod= 2, but I got the following error. 

![Hydro -Error](https://user-images.githubusercontent.com/91909348/159338945-2f9ee06a-db1a-4c23-85a3-81e96ecf95ca.PNG)

I think this makes sense for the s-function not to work with a setting of 2, because we are not using linearized (state space) models in the nonlinear model, so it should only work with both being set to 1?


Thanks, 
Qusay.

**Reply by jjonkman** [reply]
Dear @QusayHawari,

Yes, if the state-space model is a good fit, the linear and nonlinear model responses should be similar, at least for small motions.  The disturbance should not be "5 m", but a sinusoid with the same amplitude and phase as the wave elevation used in the nonlinear model (amplitude = `WaveHs`/2 = 2.5 m, period = `WaveTp`).

I'm not sure I understand your second question, but the wave direction specified via `WaveDir` should lie within the range of wave directions accounted for in the potential-flow solution.

Best regards,

**Reply by QusayHawari** [reply]
Hi,


Thank you for the clarification, much appreciated.

Regarding the second question, I think it would be less accurate to set ExctnMod = RdtnMod =2 when we want to analyze nonlinear responses as the waves and radiations will be represented in state space form (which is based on operating points). For better validation of linear and nonlinear it would be accurate to set ExctnMod = RdtnMod= 1 for the nonlinear model. Is that correct?

I think I have bugged you a lot. I will re-read the material on hydrodynamics and understand more. 

Thanks,
Qusay.

**Reply by jjonkman** [reply]
Dear @QusayHawari,

The state-space form of the potential-flow wave-excitation and wave-radiation models (`ExctnMod` = `RdtnMod` = 2) can be nearly as accurate as the DFT/convolution models (`ExctnMod` = `RdtnMod` = 1) if the state-space fit is good.  To check the state-space model fit, it would be good to compare the nonlinear time-domain solution with `ExctnMod` = `RdtnMod` = 2 against the results with `ExctnMod` = `RdtnMod` = 1; such a comparison would likely have to be done for multiple sea state conditions because the accuracy of the state-space model would likely be frequency dependent.  To verify the accuracy of the linear model, it is always useful to compare the time-domain derived responses from the linear model with the those of the nonlinear model.

Best regards,

**Reply by QusayHawari** [reply]
Hi, 

I ran the nonlinear model for two cases. One was using ExctnMod = RdtnMod = 2, and the other using ExctnMod = RdtnMod = 1. I also tested each case using two wave elevations (5 m and 10 m). I can confirm that the responses for both cases match one another having the same frequency, but with a slightly different amplitude. You can see the platform pitch response for both cases below.

![Exct_Rad_2_1_cmp](https://user-images.githubusercontent.com/91909348/159684073-fd90859e-ba11-4153-9ded-615775a6d25c.PNG)

I then went to test the s-function through Simulink using both cases (exact settings). The s-function worked fine for  ExctnMod = RdtnMod = 1, however when  ExctnMod = RdtnMod = 2, I got the error I showed you earlier:
![image](https://user-images.githubusercontent.com/91909348/159685406-bbb96af4-2a41-4edf-b303-f66095b7d4f5.png)
 

If I re-run the s-function again (keeping ExctnMod = RdtnMod = 2), I get a new error:

![MS_SFNC_ERROR](https://user-images.githubusercontent.com/91909348/159684528-ae964719-1a15-4e5c-8a07-47e335c1ab66.PNG)

Any further changes I make to the fst input files (e.g., setting ExctnMod = RdtnMod  back to 1) will result in the last error I showed. Restarting Matlab got the s-function working again (just for settings of 1).

Any idea why this error occurs with the s-function but not when compiling OpenFAST directly?

Thanks,
Qusay.

**Reply by jjonkman** [reply]
Dear @QusayHawari,

Did you compile the OpenFAST executable and the OpenFAST S-Function for Simulink yourself, or use one of the precompiled versions provided with one of the OpenFAST releases?  Are the versions of the OpenFAST executable and OpenFAST S-Function for Simulink compiled from the same version of the OpenFAST source code?

Best regards,

**Reply by QusayHawari** [reply]
Hi,


I am using OpenFAST v3.0, along with all it's precompiled tools , such as the mex64 for the s-function. 

https://github.com/OpenFAST/openfast/releases/tag/v3.0.0

Shall I upgrade to v3.1 and check?

Thanks,
Qusay.

**Reply by andrew-platt** [reply]
Dear @QusayHawari,

The error you are receiving about the summary `.HD.sum` file being locked is likely due to a bug in HydroDyn -- the file is likely not getting closed when HD fails, which is preventing it from being opened on the second call from SimuLink.  One solution for this is to set the `HDSum` flag in the HydroDyn input file to `false`. This bug may still be present in v3.1.

I don't have any good ideas on the other issue.

Regards,
Andy

**Reply by QusayHawari** [reply]
Thanks Andrew. 

I noticed that MATLAB seems to crash if I set PCMode to 4, but VSContrl to 5 when using the s-function. I assume that all control inputs should be commanded from the the s-function, but say if I wanted to only command the blade pitch while relying on the ROSCO controller to control the generator torque (setting VSContrl = 5). Is that not applicable, as MATLAB always crashes if I do that.


Thanks in advance,
Qusay.

**Reply by jjonkman** [reply]
Dear @QusayHawari,

This is probably a better question for posting on ROSCO issues: https://github.com/NREL/ROSCO/issues.  I'm not familiar enough with ROSCO to know if you can have ROSCO control generator torque, but not blade pitch.

Best regards,

**Reply by QusayHawari** [reply]
Hi @jjonkman , 

Just going back to the state space models for the wave excitation and radiation. 

In the .ssexctn file for example, there are a number of states, e.g. 8 relating to the platform surge, and 8 relating to the platform pitch, and similarly for other platform DOFs as shown below. The input to this model is one, which is the wave elevation. The outputs are six, which are the 6 DOF motions related to the input's excitation. What I am not sure about is why there are multiple states for each DOF motion. I looked through the documents you sent a while back to see why for example there are 8 platform surge states instead of just one. 

![image](https://user-images.githubusercontent.com/91909348/162029673-c73a9437-a29a-41ac-aad4-105db1476dad.png)

Similarly for the radiation state space models. There are six inputs and six outputs relating to the 6 platform DOFs, but there are multiple states defined for each platform motion. 

As an example, I took the section relating to the states for the platform surge from the full averaged state space model (after performing MBC3) linearised at 15m/s. 

A matrix for 8 states relating to the platform surge (defined in .ssexctn)
![psg_states_8](https://user-images.githubusercontent.com/91909348/162031597-c8db858c-7034-4f13-bfcb-ee567d603c4e.PNG)

B matrix, which is the wave elevation input.
![psg_inp_8](https://user-images.githubusercontent.com/91909348/162031610-c7422cb0-e661-4bdd-8d79-75dee57b32a6.PNG)

All states have significant value and none of them can be eliminated, and it's difficult to understand what those states represent.

Thanks,
Qusay

**Reply by jjonkman** [reply]
Dear @QusayHawari,

The hydrodynamic states do not have a physical interpretation.  As I mentioned in a response above, the states capture the frequency dependence and memory effects and the papers I reference should hopefully provide clarity.  

And to correct a couple comments you made:

- In the state-space wave-radiation model, the inputs are the 6 body velocities (3 translational and 3 rotational velocities) and the outputs are the 6 wave-radiation loads (3 forces, 3 moments).  The states capture the memory effect between velocity and load.
- In the state-space wave-excitation model, the input is the wave elevation and the outputs are the 6 wave-excitation loads (3 forces, 3 moments).  The states capture the frequency-dependence between wave elevation and load.

Best regards,

**Comment by Melancho7ee** [comment]
Sorry to bother, but could you please send me an IEA 15MW WindCrete model. @QusayHawari 
I am interested in the IEA 15MW WindCrete Model and downloaded it from UPC - WINDCRETE OpenFAST model 15MW FOWT - Grand Canary Island | Zenodo. the website where I download WindCrete is https://zenodo.org/record/4322446#.ZF4t2xFBzNp
However, I cannot run it with OpenFAST/Simulink.
I checked the files such as *.fst and AeroDyn15.dat and found that the files are not consistent with OpenFAST v2.3.0, 2.4.0, 3.0.0 or 3.4.1 (Some lines are missing comapared with "IEA-15-240-RWT-UMaineSemi.fst" which is downloadable from GitHub).
I am wondering which version of OpenFAST it is. Actually I think the website provided a wrong or incomplete version of WindCrete model
So Does anyone how to get a model that can run with OpenFAST+Simulink directly just like the NREL 5MW models?? Many Thanks

**Reply by QusayHawari** [reply]
Hi,

You need to update the parameters relating to the WindCrete floating platform in HydroDyn, MoorDyn, ElastoDyn, and SubDyn to match those for the WindCrete platform. You can use the input files and OpenFAST version you currently have, but update the platform data to match WindCrete. Please check the discussion here: https://github.com/IEAWindTask37/IEA-15-240-RWT/issues/103#issue-1413393225

Qusay.

**Reply by Melancho7ee** [reply]
Okayokay. Many Thanks! I will try again

**Reply by Melancho7ee** [reply]
Sorry to bother, but after i modified my input files in openfast 3.4.1, i received following error
`FAST_InitializeAll:InitModuleMappings:ED_2_AD_TowerMotion:MeshMapCreate:CreateMotionMap_L2_to_L2:CreateMapping_ProjectToLine2:Node 10 does not project onto any line2 element. Closest distance is 3.9362 m.`
What is wrong?

**Reply by jjonkman** [reply]
Dear @Melancho7ee,

A similar question was asked and answered in the following discussion: https://github.com/OpenFAST/openfast/discussions/908.

Best regards,
